include/seqan3/alignment/matrix/detail/aligned_sequence_builder.hpp

```diff
@@ -0,0 +1,249 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::detail::aligned_sequence_builder.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <type_traits>
+#include <vector>
+
+#include <seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp>
+#include <seqan3/alignment/matrix/detail/matrix_coordinate.hpp>
+#include <seqan3/alignment/matrix/trace_directions.hpp>
+#include <seqan3/alphabet/gap/gapped.hpp>
+#include <seqan3/range/decorator/gap_decorator.hpp>
+#include <seqan3/range/view/slice.hpp>
+#include <seqan3/range/view/view_all.hpp>
+#include <seqan3/std/concepts>
+#include <seqan3/std/ranges>
+
+namespace seqan3::detail
+{
+
+/*!\brief Transforms the given range types into their aligned sequence format.
+ * \ingroup alignment_matrix
+ * \tparam fst_rng_t The type of the first range to be transformed to a aligned sequence type.
+ * \tparam sec_rng_t The type of the second range to be transformed to a aligned sequence type.
+ *
+ * \details
+ *
+ * First this traits refines the original range types with seqan3::view::slice. Then it transforms the given range type
+ * into a seqan3::gap_decorator if the expression is not ill-formed, otherwise the transformed aligned sequence type
+ * is a std::vector over the respective alphabet decorated with seqan3::gapped.
+ */
+template <std::ranges::ViewableRange fst_rng_t,
+          std::ranges::ViewableRange sec_rng_t>
+struct aligned_sequence_builder_default_traits
+{
+private:
+    //!\brief Helper type alias for sliced range.
+    using fst_slice_t = decltype(std::declval<fst_rng_t>() | view::slice(0, 1));
+    //!\copydoc aligned_sequence_builder_default_traits::fst_slice_t
+    using sec_slice_t = decltype(std::declval<sec_rng_t>() | view::slice(0, 1));
+
+    //!\brief Helper function to determine the transformed type.
+    template <typename rng_t>
+    //!\cond
+        requires std::Constructible<gap_decorator<rng_t>>
+    //!\endcond
+    static auto aligned_type(rng_t *) { return gap_decorator<rng_t>{}; }
+
+    //!\overload
+    template <typename rng_t>
+    static auto aligned_type(rng_t *) { return std::vector<gapped<value_type_t<rng_t>>>{}; }
+
+public:
+    //!\brief The aligned sequence type for the first sequence.
+    using fst_aligned_t = decltype(aligned_type(static_cast<fst_slice_t *>(nullptr)));
```

Comment:
Well, it I used a similar approach as used in ranges library. declval would also be ok since it is in an unevaluated context but I guess I revert it back to just pass the types as template arguments

Reply:
You can, but you don't need to.